File: strings/rabin_karp.go
package main

import "fmt"

const alphabetSize = 256
const modulus = 1000003

func RabinKarp(pattern, text string) bool {
	pLen := len(pattern)
	tLen := len(text)

	if pLen > tLen {
		return false
	}

	var pHash, textHash, modulusPower int64 = 0, 0, 1

	for i := 0; i < pLen; i++ {
		pHash = (int64(pattern[i]) + pHash*alphabetSize) % modulus
		textHash = (int64(text[i]) + textHash*alphabetSize) % modulus
		if i < pLen-1 {
			modulusPower = (modulusPower * alphabetSize) % modulus
		}
	}

	for i := 0; i <= tLen-pLen; i++ {
		if pHash == textHash {
			if text[i:i+pLen] == pattern {
				return true
			}
		}

		if i < tLen-pLen {
			textHash = (textHash - int64(text[i])*modulusPower) * alphabetSize
			textHash += int64(text[i+pLen])
			textHash %= modulus
			if textHash < 0 {
				textHash += modulus
			}
		}
	}
	return false
}

func main() {
	text := "ABABDABACDABABCABAB"
	pattern := "ABABCABAB"
	
	if RabinKarp(pattern, text) {
		fmt.Println("Pattern found!")
	} else {
		fmt.Println("Pattern not found.")
	}
}

File: strings/rabin_karp.js
/**
 * Rabin-Karp Algorithm
 */
const ALPHABET_SIZE = 256;
const MODULUS = 1000003;

function rabinKarp(pattern, text) {
    const pLen = pattern.length;
    const tLen = text.length;

    if (pLen > tLen) return false;

    let pHash = 0;
    let textHash = 0;
    let modulusPower = 1;

    for (let i = 0; i < pLen; i++) {
        pHash = (pattern.charCodeAt(i) + pHash * ALPHABET_SIZE) % MODULUS;
        textHash = (text.charCodeAt(i) + textHash * ALPHABET_SIZE) % MODULUS;
        if (i < pLen - 1) {
            modulusPower = (modulusPower * ALPHABET_SIZE) % MODULUS;
        }
    }

    for (let i = 0; i <= tLen - pLen; i++) {
        if (pHash === textHash) {
            if (text.substring(i, i + pLen) === pattern) {
                return true;
            }
        }

        if (i < tLen - pLen) {
            textHash = (textHash - text.charCodeAt(i) * modulusPower) * ALPHABET_SIZE + text.charCodeAt(i + pLen);
            textHash %= MODULUS;
            if (textHash < 0) textHash += MODULUS;
        }
    }
    return false;
}

const text = "ABABDABACDABABCABAB";
const pattern = "ABABCABAB";

if (rabinKarp(pattern, text)) {
    console.log("Pattern found!");
} else {
    console.log("Pattern not found.");
}
